service/history/timerQueueProcessorBase.go

```diff
@@ -632,50 +634,94 @@ func (t *timerQueueProcessorBase) deleteWorkflow(task *persistence.TimerTaskInfo
 	return nil
 }
 
-func (t *timerQueueProcessorBase) archiveWorkflow(task *persistence.TimerTaskInfo, msBuilder mutableState, context workflowExecutionContext) error {
+func (t *timerQueueProcessorBase) archiveWorkflow(task *persistence.TimerTaskInfo, msBuilder mutableState, workflowContext workflowExecutionContext) error {
 	domainCacheEntry, err := t.historyService.shard.GetDomainCache().GetDomainByID(task.DomainID)
 	if err != nil {
 		return err
 	}
 
-	// TODO ycyang: rewrite archiveRequest
-	req := &archiver.ArchiveRequest{
-		ShardID:              t.shard.GetShardID(),
-		DomainID:             task.DomainID,
-		DomainName:           domainCacheEntry.GetInfo().Name,
-		WorkflowID:           task.WorkflowID,
-		RunID:                task.RunID,
-		EventStoreVersion:    msBuilder.GetEventStoreVersion(),
-		BranchToken:          msBuilder.GetCurrentBranch(),
-		NextEventID:          msBuilder.GetNextEventID(),
-		CloseFailoverVersion: msBuilder.GetLastWriteVersion(),
-		BucketName:           domainCacheEntry.GetConfig().HistoryArchivalURI,
-	}
-
-	// send signal before deleting mutable state to make sure archival is idempotent
-	if err := t.historyService.archivalClient.Archive(req); err != nil {
-		t.logger.Error("failed to initiate archival", tag.Error(err),
-			tag.WorkflowID(task.WorkflowID),
-			tag.WorkflowRunID(task.RunID),
-			tag.WorkflowDomainID(task.DomainID),
-			tag.ShardID(t.shard.GetShardID()),
-			tag.TaskID(task.GetTaskID()),
-			tag.FailoverVersion(task.GetVersion()),
-			tag.TaskType(task.GetTaskType()))
+	executionStats, err := workflowContext.loadExecutionStats()
+	if err != nil {
 		return err
 	}
+	archiveInline := executionStats.HistorySize < int64(t.config.TimerProcessorHistoryArchivalSizeLimit())
+	taggedLogger := t.logger.WithTags(
+		tag.Error(err),
+		tag.WorkflowID(task.WorkflowID),
+		tag.WorkflowRunID(task.RunID),
+		tag.WorkflowDomainID(task.DomainID),
+		tag.ShardID(t.shard.GetShardID()),
+		tag.TaskID(task.GetTaskID()),
+		tag.FailoverVersion(task.GetVersion()),
+		tag.TaskType(task.GetTaskType()),
+	)
+	if !archiveInline {
+		req := &archiver.ArchiveRequest{
+			ShardID:              t.shard.GetShardID(),
+			DomainID:             task.DomainID,
+			DomainName:           domainCacheEntry.GetInfo().Name,
+			WorkflowID:           task.WorkflowID,
+			RunID:                task.RunID,
+			EventStoreVersion:    msBuilder.GetEventStoreVersion(),
+			BranchToken:          msBuilder.GetCurrentBranch(),
+			NextEventID:          msBuilder.GetNextEventID(),
+			CloseFailoverVersion: msBuilder.GetLastWriteVersion(),
+			URI:                  domainCacheEntry.GetConfig().HistoryArchivalURI,
+		}
+
+		// send signal before deleting mutable state to make sure archival is idempotent
+		if err := t.historyService.archivalClient.Archive(req); err != nil {
```

Comment:
What do you think about moving both cases into the archivalClient? That way when you implement DB scanner you do not have to duplicate this code?

Reply:
Good point! Updated.